documentation/source/protocol_extension.rst

```diff
@@ -0,0 +1,90 @@
+How to write a protocol extension
+=================================
+
+Cocotb take care of the core communication between the simulator and testbench.
+To use some known busses like wishbone, avalon, spi, i2c,... An extension is required.
+
+This is a guideline that will explain how to write a python module protocol
+extension for cocotb.
+
+Naming convention
+-----------------
+
+A cocotb extension module must begin by `cocotb_` and terminate with protocol
+name.
+
+For exemple, with spi the module extension is `cocotb_spi`.
+
+Bus driver
+----------
+
+A Bus driver is a python class that factorize common behavior and check for
+presence of some signals.
+
+A Bus driver is an active class that be driven in the main testbench. Example is
+given with avalon bus in `cocotb/drivers/avalon.py`.
+
+With AvalonMaster class we can generate stimulis for reading and writing on bus.
+
+The main class is named `BusDriver` and is defined in `cocotb/drivers/__init__.py`
+
+TODO: give an example.
+
+Bus Monitor
+-----------
+
+A Bus monitor is mainly a passive class that will monitor the bus.
+
+TODO: explain monitor
+
+Python module
+-------------
+
+The module file hierarchy should be as following ::
+
+  ├── cocotb_spi/
+  │   └── __init__.py
+  ├── README.md
+  └── setup.py
+
+With python module source code in directory cocotb_spi and module configuration
+in setup.py
+
+setup.py is a python configuration file that be used for packaging. An example
+is given above ::
+
+  try:
+    from setuptools import setup
+  except ImportError:
+    from distutils.core import setup
+
+  setup(name='cocotb_spi',
```

Comment:
Should the package be called `cocotb-spi`? We have few in such convention.

Reply:
Yes, '-' is better I think. I like the eric proposition with `cocotb.ext.spi` but is it possible with an external package ?

Reply:
This is an entirely different question - there are two things to decide here:

* The module name, as in `import module_name`. This is what goes in `py_modules`, and the name of the python file
* The distribution name, as in `pip install distribution-name`. This is what goes here in setup.py

@themperek is arguing that we don't need to choose these two names to be the same. While this is true, users might find it easier to realize that `import cocotb_spi` means they need to run `pip install cocotb_spi` instead of `pip install cocotb-spi`, so it might not be worth making the names different.

I'll see if I can get anywhere with an example project using `cocotb.ext.spi`.

Reply:
@wallento use `cocomod` as prefix for his [protocol package](https://github.com/wallento/cocomod-uart). Is there other package that using this `cocomod` name or is it only name from @wallento packages ?

Reply:
And yes, it should use hypen ('-') for the package name

Reply:
Ah, sorry, in sphinx its sphinxcontrib, sphinx.ext is in the source tree too

Reply:
@eric [you can't extend from the cocotb namespace](https://packaging.python.org/guides/packaging-namespace-packages/#pkgutil-style-namespace-packages). We have actually tried this recently.